packages/tasit-action/src/TasitAction.test.js

```diff
@@ -84,109 +77,292 @@ describe("Contract", () => {
+        errorFn();
+      });
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      expect(confirmationFn.called).to.be.true;
+
+      await new Promise(resolve => setTimeout(resolve, 3000));
+
+      expect(
+        subscription.listenerCount("confirmation"),
+        "listener should be removed after timeout"
+      ).to.be.equals(0);
+
+      expect(errorFn.called).to.be.true;
+    });
+
+    it("should emit error event when orphan/uncle block occurs", async () => {
+      subscription = simpleStorage.setValue("hello world");
 
-        const value = await simpleStorage.getValue();
-        expect(value).to.equal(rand);
+      const confirmationFn = sinon.fake();
+      const errorFn = sinon.fake();
 
-        // UnhandledPromiseRejectionWarning
+      const callback = message => {
+        confirmationFn();
+      };
+
+      await subscription.on("confirmation", callback);
+
+      await subscription.on("error", error => {
+        errorFn();
+        subscription.removeAllListeners();
+
+        // FIXME: Assertion not working
         //expect(1).to.equal(2);
-      }
-    };
+        expect(error.message).to.match(/uncle/);
+      });
+
+      const snapshotId = await createSnapshot(simpleStorage.getProvider());
+
+      await mineBlocks(simpleStorage.getProvider(), 15);
 
-    subscription.on("confirmation", onMessage);
+      expect(confirmationFn.called).to.be.true;
 
-    await mineBlocks(simpleStorage.getProvider(), 8);
+      await revertFromSnapshot(simpleStorage.getProvider(), snapshotId);
+
+      await mineBlocks(simpleStorage.getProvider(), 20);
+
+      // Broadcast same message again (simulate reorg)
+      subscription = simpleStorage.setValue("hello world");
+
+      expect(errorFn.called).to.be.true;
+    });
   });
 
-  it("should throw error when subscribing on invalid event", async () => {
-    const events = ["ValueChanged", "InvalidEvent"];
-    try {
+  describe("contract events subscriptions tests", async () => {
+    beforeEach("assign a wallet to the contract", () => {
+      expect(() => {
+        simpleStorage.setWallet(wallet);
+      }).to.not.throw();
+    });
+
+    it("should throw error when subscribing on invalid event", async () => {
+      const events = ["ValueChanged", "InvalidEvent"];
+
+      expect(() => {
+        subscription = simpleStorage.subscribe(events);
+      }).to.throw();
+    });
+
+    it("should throw error then listening on invalid event", async () => {
+      const events = ["ValueChanged"];
       const subscription = simpleStorage.subscribe(events);
-      assert(false, "subscription with invalid event");
-    } catch (e) {
-      assert(true);
-    }
-  });
 
-  it("should throw error then listening on invalid event", async () => {
-    const events = ["ValueChanged"];
-    const subscription = simpleStorage.subscribe(events);
+      expect(() => {
+        subscription.on("InvalidEvent", () => {});
+      }).to.throw();
```

Comment:
Do we want to pick a policy on whether expecting any error is sufficient? Should we expect a specific error message too?

Reply:
Perfect. I'll resolve this thread here.